strings/base_macros.h

```diff
@@ -93,19 +132,23 @@ namespace winrt::impl
     // have heavy binary size overhead when templates cause many permutations to exist.
     struct slim_source_location
     {
-        [[nodiscard]] static consteval slim_source_location current(
-            const std::uint_least32_t line = __builtin_LINE(),
-            const char* const file = __builtin_FILE()) noexcept
+        [[nodiscard]] static WINRT_IMPL_CONSTEVAL slim_source_location current(
```

Comment:
As best as I can tell this approach resolves to the correct thing because the preprocessor is doing the work.  Even translation units in the same library are compatible with different flags.  I added another test case to my little app and it has the expected output:

(This code is in the cpp20 static lib alongside the previously shared sample)
```cpp
```
#define WINRT_NO_SOURCE_LOCATION
#include <winrt/base.h>

void Cpp20StaticLibSuppressSourceLocation()
{
    try
    {
        winrt::check_hresult(E_FAIL);
    }
    catch (const winrt::hresult_error& e)
    {
        wprintf(L"C++20 Static Lib (WINRT_NO_SOURCE_LOCATION): %ws\n", e.message().c_str());
    }
}
```
winrt::hresult_error thrown at line 8 in V:\T\davidmac.randomprojects\CppWinrtMixCpp17AndCpp20\Cpp20StaticLib\Cpp20StaticLib.cpp in Cpp20StaticLib with error code 0x80004005
winrt::hresult_error thrown at line 8 in V:\T\davidmac.randomprojects\CppWinrtMixCpp17AndCpp20\Cpp20StaticLib\Cpp20StaticLib.cpp in Cpp20StaticLib with error code 0x80004005
C++20 Static Lib: Unspecified error
winrt::hresult_error thrown at line 0 in (null) in (null) with error code 0x80004005
winrt::hresult_error thrown at line 0 in (null) in (null) with error code 0x80004005
C++20 Static Lib (WINRT_NO_SOURCE_LOCATION): Unspecified error
```

Reply:
The C++ standard says ...

> There can be more than one definition in a program of [...] inline function [...] as long as all of the following is true:
> * each definition appears in a different translation unit
> * each definition consists of the same sequence of tokens
> * [... more rules]
> * the rules above apply to every default argument used in each definition

My understanding is that by conditionally changing the *default* arguments in the definition of `slim_source_location::current()`, we are introducing an ODR violation on that function - there are two definitions of the function with different default argument values.

It might be a relatively tame violation, because we're not changing the type of the params, nor which params have default values. In practice, it's probably common that we see those args evaluated using the declaration currently visible to that translation unit. But in the presence of things like whole-program optimization, I don't know for certain if the evaluation of default args can be deferred to link time, at which point we can't predict which version is used.

Reply:
I tried your suggestion of replacing `check_hresult` with `winrt::Windows::Foundation::Uri(L"InvalidUri");` and that does not behave well.  The cpp17 library wins the linker battle and the source location is always empty.

Reply:
`current` doesn't violate ODR. The default arguments are considered to be separate definitions from the function definition. The function definition itself is the same in all TUs (after the change, to be clear) regardless of support for `source_location` or the definition of `WINRT_NO_SOURCE_LOCATION`. The difference in the default argument definitions _does_ affect the definitions of other functions that call `slim_source_location::current` in ways that use those default arguments. The effect is as if the tokens in the definition of the default argument appeared in the definition of the calling function (https://eel.is/c++draft/basic.def.odr#15.11). If some of those callers are defined in multiple TUs that differ on support for `source_location` or the definition of `WINRT_NO_SOURCE_LOCATION`, then those callers are violating the ODR.

Outside of the guarantees the Standard provides, I'll say this seems like a fairly benign ODR violation. I don't think it's unreasonable to remove the `detect_mismatch`, but even if you're incredibly conservative I'd suggest providing an escape hatch macro for users to define to disable the mismatch detection.

Reply:
I think your assessment is in alignment with what I'm seeing in a local test app.

There is an ODR violation but it is not a dangerous one.  For a given `winrt::consume_X` generated function that has confliction definitions one will include source, the other will not.  The linker then chooses which is the winner and the resulting binary will either have, or not have, source information in that function based on the choice.  Regardless of the choice the `slim_source_location` class is the same and the ABI is the same too so there is no corruption or anything really bad like that.

I think the matrix is roughly as follows:
1) No ODR, everything is consistent.  This is the happy option.
2) Mixing source location options.  For every function that is actually called in a given translation unit:
2a) Linker chooses the one with source information.  Final binary has source info for that call no matter which compiland is calling it.
2b) Linker chooses the one without source information.  Finaly binary does not have source info for that call, no matter which compiland is calling it.
3) Functions that are only called from one compiland or the other will have the source information for that compiland. In other words:
3a) Only cpp17 code calls a given function.  That function will not have source info.
3b) Only cpp20 code calls a given function.  That function will have source info.

Overall this feels fine to me.  We really shouldn't be mixing static libs across language versions so in most cases this doesn't matter.  When it does matter, the source info may be missing in some cases, but at no point will this introduce crashes or other major problems.